packages/cosmic-proto/test/address-hooks.test.js

```diff
@@ -0,0 +1,214 @@
+import rawTest from '@endo/ses-ava/prepare-endo.js';
+
+import bundleSourceAmbient from '@endo/bundle-source';
+import { importBundle } from '@endo/import-bundle';
+
+/**
+ * @type {import('ava').TestFn<{
+ *   addressHooks: import('../src/address-hooks.js');
+ * }>}
+ */
+const test = rawTest;
+
+const makeTestContext = async () => {
+  const bundleSource = bundleSourceAmbient;
+  const loadBundle = async specifier => {
+    const modulePath = new URL(specifier, import.meta.url).pathname;
+    const bundle = await bundleSource(modulePath);
+    return bundle;
+  };
+
+  const evaluateBundle = async (bundle, endowments = {}) => {
+    return importBundle(bundle, endowments);
+  };
+
+  const importSpecifier = async (specifier, endowments = {}) => {
+    const bundle = await loadBundle(specifier);
+    return evaluateBundle(bundle, endowments);
+  };
+
+  const addressHooks = await importSpecifier('../src/address-hooks.js');
+
+  return { addressHooks };
+};
+
+test.before(async t => {
+  t.context = await makeTestContext();
+});
+
+/**
+ * @type {import('ava').Macro<
+ *   [string, ArrayLike<number> | undefined, ArrayLike<number>, string],
+ *   { addressHooks: import('../src/address-hooks.js') }
+ * >}
+ */
+const roundtripMacro = test.macro({
+  title(providedTitle = '', prefix, addrBytes, hookData) {
+    const space = providedTitle.endsWith(' ') ? '' : ' ';
+    return `${providedTitle}${space}prefix: ${prefix}, addrBytes: ${addrBytes}, hookData: ${hookData}`;
+  },
+  exec(t, prefix, addrBytes, hookData, expected) {
+    const { encodeBech32, joinHookedAddress, splitHookedAddress } =
+      t.context.addressHooks;
+    const baseAddress = encodeBech32(prefix, addrBytes || []);
+    const encoded = joinHookedAddress(baseAddress, hookData);
+    t.deepEqual(encoded, expected);
+    const decoded = splitHookedAddress(encoded);
+    t.deepEqual(decoded, {
+      baseAddress,
+      hookData: new Uint8Array(hookData || []),
+    });
+  },
+});
+
+test(
+  'roundtripEmpty',
```

Comment:
after mulling over what this test establishes, I thought about property testing / fuzzing to be sure the encoding is unambiguous - that no 2 distinct inputs lead to the same encoded address.

But that would say that the design is flawed. And as noted above, I assume the design is agreed by the relevant parties.